### ISSUE TITLE
Turn on super parameters in the flutter repository

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#100575]
New language feature in Dart 2.17: https://github.com/dart-lang/language/issues/1855

There's a lint for using it when appropriate: `use_super_parameters`

- [x] Migrate all code in `flutter`
  - [x] https://github.com/flutter/flutter/pull/100515
  - [x] https://github.com/flutter/flutter/pull/100516
  - [x] https://github.com/flutter/flutter/pull/100509
  - [x] https://github.com/flutter/flutter/pull/100905
  - [x] https://github.com/flutter/flutter/pull/100510
  - [x] https://github.com/flutter/flutter/pull/100514
  - [x] https://github.com/flutter/flutter/pull/101065
  - [x] https://github.com/flutter/flutter/pull/101073
- [x] https://github.com/flutter/flutter/pull/101074
- [x] https://github.com/flutter/flutter/pull/101157
- [x] Suggest IDEs to use super params in macros:
  - [x] https://github.com/Dart-Code/Dart-Code/issues/3899
  - [x] https://github.com/flutter/flutter-intellij/issues/6071 
- [x] Suggest lint for inclusion in `package:lints`: https://github.com/dart-lang/lints/issues/75
- [x] For later: https://github.com/flutter/flutter/issues/101068

**Comment by goderbauer** [comment]
The autofix for this has a bug (https://github.com/dart-lang/sdk/issues/48633), which currently makes it cumbersome to fix this for the framework itself. Waiting for the pending fix to roll in.

**Comment by goderbauer** [comment]
There's also a bug with non-const subclasses of const superclasses (see https://github.com/dart-lang/sdk/issues/48642), which is blocking most of the PRs linked above.

**Comment by mit-mit** [comment]
@goderbauer both those two Dart SDK issues are now resolved!

**Comment by goderbauer** [comment]
Awesome! Looks like the fix for https://github.com/dart-lang/sdk/issues/48642 hasn't rolled into Flutter yet. As soon as it does, I will give this another go.

**Comment by pq** [comment]
@goderbauer: definitely let me know if you hit any snags.  Hopefully we're good to go! :🤞

**Comment by goderbauer** [comment]
I am a little hesitant to migrate the sample code we show in our API docs (https://api.flutter.dev/) over just yet. Lots of developers with existing apps copy code from there and if we migrate over to super params it wouldn't work in their existing apps until they upgrade their SDK constraints to 2.17. Unfortunately, the message they get if they have the incorrect bounds is not super actionable, `flutter analyze .` just says 

```
error • This requires the 'super-parameters' language feature to be enabled • lib/main.dart:9:12 • experiment_not_enabled
```

So, long story short: I think we should migrate all code over except for the API samples for now. Once the super parameter feature is released and a little better known we should follow up and migrate the sample code as well.

WDYT @mit-mit @pq @gspencergoog

**Comment by pq** [comment]
> Unfortunately, the message they get if they have the incorrect bounds is not super actionable, flutter analyze .

Yeah. This is a really unfortunate message.  We should be able to improve it.

/fyi @bwilkerson @stereotype441 

>  I think we should migrate all code over except for the API samples for now. 

SGTM.

> Once the super parameter feature is released and a little better known we should follow up and migrate the sample code as well.

Sounds good, though a little unfortunate that it's another thing someone needs to track and just grows the tail of language feature shipping a wee bit longer.

(Aside: @leafpetersen, @devoncarew (or anyone): I wonder if we shouldn't add some kind of a follow-up task to the language feature tracking template for documentation that we want to lag intentionally?)

**Comment by leafpetersen** [comment]
@itsjustkevin is probably the person to contact about getting something into the release checklist.

**Comment by bwilkerson** [comment]
That message is defined in the `front_end` package. I'll just note that it has a correction message associated with it:

> Try updating your pubspec.yaml to set the minimum SDK constraint to #string2 or higher, and running 'pub get'.

Is it the case that `flutter analyze` doesn't display the correction message?

**Comment by goderbauer** [comment]
> Is it the case that flutter analyze doesn't display the correction message?

`flutter analyze` does not show that correction message. I just checked, I can see it in the IDE, though, if I hover over the error message (not super-discoverable...).

**Comment by bwilkerson** [comment]
> `flutter analyze` does not show that correction message.

We might want to change that. `dart analyze` prints both the correction message and a link to documentation (if it exists), and I think both are useful to users.

> I can see it in the IDE ... not super-discoverable...

It's more visible in the hover than in the "Dart Analysis" view. It's also more visible in VS Code's Problems view.

**Comment by goderbauer** [comment]
> We might want to change that. dart analyze prints both the correction message and a link to documentation (if it exists), and I think both are useful to users.

Filed https://github.com/flutter/flutter/issues/100932 to track this.

**Comment by goderbauer** [comment]
Getting the PRs submitted for this is currently blocked on some infra issues: https://github.com/flutter/flutter/issues/100731

**Comment by gspencergoog** [comment]
> So, long story short: I think we should migrate all code over except for the API samples for now.

Works for me.

**Comment by mit-mit** [comment]
> So, long story short: I think we should migrate all code over except for the API samples for now.

SGTM

**Comment by goderbauer** [comment]
I am running into another problem while migrating the framework. Unfortunately, I wasn't able to reproduce this outside of the framework. See https://github.com/flutter/flutter/pull/100905/files#r837961405 for details.

**Comment by goderbauer** [comment]
The migration to super parameters has also caused a failure in one of the Dart CI bots: https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8818330122470647153/+/u/test_results/new_test_failures__logs_

Unfortunately, I am unfamiliar with this bot and don't know how to fix this. The test runs just fine when executed directly in the flutter/flutter repository. I suspect that the CI bot doesn't honor the pubspec.yaml for the test and therefore doesn't enable the experiment. Somewhere on the bot we need to bump the lower sdk bound to 2.17.0-0, but I don't know where to do that since I am unfamiliar with the bot. @mit-mit Do you happen to know who could help out here?

/cc @brianquinlan

**Comment by goderbauer** [comment]
> I am running into another problem while migrating the framework. Unfortunately, I wasn't able to reproduce this outside of the framework. See https://github.com/flutter/flutter/pull/100905/files#r837961405 for details.

I was able to reproduce this one and filed https://github.com/dart-lang/sdk/issues/48697 with details. It is blocking further migration of the framework.

**Comment by goderbauer** [comment]
Another minor issue with the autofix for this lint (not blocking): https://github.com/dart-lang/sdk/issues/48698

**Comment by goderbauer** [comment]
Also, the analyzer could be smarter about this: https://github.com/dart-lang/sdk/issues/48699

**Comment by goderbauer** [comment]
A test on the migration PR for flutter_tools (https://github.com/flutter/flutter/pull/100509) is failing in an odd way:

```
$ dart run test test/general.shard/project_file_invalidator_test.dart                                                                                 
00:00 +0: loading test/general.shard/project_file_invalidator_test.dart                                                                                                                                  Unhandled exception:
Invalid argument(s): Type parameter TypeParameter(MultiRootFileSystemEntity.D) is not indexed
#0      TypeParameterIndexer.[] (package:kernel/binary/ast_to_binary.dart:2927:8)
#1      BinaryPrinter.visitTypeParameterType (package:kernel/binary/ast_to_binary.dart:2432:38)
#2      TypeParameterType.accept (package:kernel/ast.dart:11915:42)
#3      BinaryPrinter.writeNode (package:kernel/binary/ast_to_binary.dart:430:10)
#4      BinaryPrinter.writeVariableDeclaration (package:kernel/binary/ast_to_binary.dart:2271:5)
#5      BinaryPrinter.writeList (package:kernel/binary/ast_to_binary.dart:331:16)
#6      BinaryPrinter.writeVariableDeclarationList (package:kernel/binary/ast_to_binary.dart:2279:5)
#7      BinaryPrinter.visitFunctionNode (package:kernel/binary/ast_to_binary.dart:1484:5)
#8      FunctionNode.accept (package:kernel/ast.dart:3871:38)
#9      BinaryPrinter.writeFunctionNode (package:kernel/binary/ast_to_binary.dart:437:10)
#10     BinaryPrinter.visitConstructor (package:kernel/binary/ast_to_binary.dart:1254:5)
#11     Constructor.accept (package:kernel/ast.dart:2488:40)
#12     BinaryPrinter.writeConstructorNode (package:kernel/binary/ast_to_binary.dart:486:10)
#13     BinaryPrinter.writeConstructorNodeList (package:kernel/binary/ast_to_binary.dart:385:7)
#14     BinaryPrinter.visitClass (package:kernel/binary/ast_to_binary.dart:1220:5)
#15     Class.accept (package:kernel/ast.dart:1433:38)
#16     BinaryPrinter.writeClassNode (package:kernel/binary/ast_to_binary.dart:472:10)
#17     BinaryPrinter.writeClassNodeList (package:kernel/binary/ast_to_binary.dart:367:7)
#18     BinaryPrinter.visitLibrary (package:kernel/binary/ast_to_binary.dart:1058:5)
#19     Library.accept (package:kernel/ast.dart:565:38)
#20     BinaryPrinter.writeLibraryNode (package:kernel/binary/ast_to_binary.dart:451:10)
#21     BinaryPrinter.writeLibraries (package:kernel/binary/ast_to_binary.dart:756:9)
#22     BinaryPrinter.writeComponentFile.<anonymous closure> (package:kernel/binary/ast_to_binary.dart:595:7)
#23     Timeline.timeSync (dart:developer/timeline.dart:157:22)
#24     BinaryPrinter.writeComponentFile (package:kernel/binary/ast_to_binary.dart:577:14)
#25     IncrementalSerializer.serialize (package:front_end/src/fasta/incremental_serializer.dart:308:13)
#26     IncrementalSerializer.writePackagesToSinkAndTrimComponent (package:front_end/src/fasta/incremental_serializer.dart:158:24)
#27     FrontendCompiler.writeDillFile (file:///opt/s/w/ir/cache/builder/sdk/pkg/frontend_server/lib/frontend_server.dart:727:29)
#28     FrontendCompiler.compile (file:///opt/s/w/ir/cache/builder/sdk/pkg/frontend_server/lib/frontend_server.dart:604:13)
<asynchronous suspension>
#29     listenAndCompile.<anonymous closure> (file:///opt/s/w/ir/cache/builder/sdk/pkg/frontend_server/lib/frontend_server.dart:1198:11)
<asynchronous suspension>
```

This reproduces locally for that test and will need some more investigation.

**Comment by mit-mit** [comment]
@goderbauer did you file an issue for that last test issue? I tried to recreate it locally, but failed.

**Comment by goderbauer** [comment]
> did you file an issue for that last test issue? I tried to recreate it locally, but failed.

Just filed https://github.com/dart-lang/sdk/issues/48708. It repos consistently for me with the PR https://github.com/flutter/flutter/pull/100509. I have not been able to extract this into a smaller repo case, though :/

**Comment by goderbauer** [comment]
Rolling the code that uses super params into google3 is currently failing because the feature is not enabled in google3. Trying to figure out if it can easily be enabled there...

**Comment by goderbauer** [comment]
Google3 issue has been resolved. A fix to the dart sdk simply hadn't rolled into google3 yet that caused the failures. Fix has rolled in now.

**Comment by goderbauer** [comment]
We are getting so close! After the last fix rolled into the framework, I found one more case where the analyzer appears to be complaining about valid code: https://github.com/dart-lang/sdk/issues/48727. I am fairly confident that that's the last issue blocking the final migration PR (https://github.com/flutter/flutter/pull/100905).

**Comment by goderbauer** [comment]
Waiting on a Dart roll which contains the fix for https://github.com/dart-lang/sdk/issues/48727. Dart rolls appear to be currently blocked on https://github.com/flutter/flutter/issues/101242 (and also https://github.com/flutter/flutter/issues/101241 as well as https://github.com/flutter/flutter/issues/101250).

**Comment by goderbauer** [comment]
The dart roll with the final fix is now blocked on https://github.com/flutter/flutter/issues/101514.

**Comment by goderbauer** [comment]
I spoke too soon. I found another issue: https://github.com/dart-lang/sdk/issues/48769.

**Comment by goderbauer** [comment]
The `package:analyzer` version used by flutter (currently 3.2.0) also errors when parsing a file with super parameters enabled. I verified, this bug is no longer present in the latest version of the analyzer (3.4.1). To upgrade to it, we also need to bump `_fe_analyzer_shared` to 38.0.0. Unfortunately, that version is not supported by dwds. I filed https://github.com/dart-lang/webdev/issues/1572 to get dwds upgraded.

**Comment by annagrin** [comment]
Released dwds 13.1.0 (bumps `_fe_analyzer_shared to ^38.0.0.)

**Comment by goderbauer** [comment]
Bumping dwds (and by extension the analyzer) has resolved the analyzer problem described in my previous problem.

Now, I am just waiting on the fix for https://github.com/dart-lang/sdk/issues/48769 to be reviewed, submitted, and rolled into the framework.

**Comment by goderbauer** [comment]
The fix for https://github.com/dart-lang/sdk/issues/48769 has rolled into the framework and works as expected. The migrated framework code now compiles without errors.

However, tests started to fail, which is caused by an incorrect migration from `dart fix`. I have filed https://github.com/dart-lang/sdk/issues/48803 for that issue. The correct fix is producing a false lint warning, for which I have filed https://github.com/dart-lang/linter/issues/3349. I believe none of these issues are blocking for the migration as there are only a hand full of cases in the framework where this appears to be a problem. I am fixing those by hand and ignoring the false lint warning for now.

**Comment by goderbauer** [comment]
I just submitted the last PR that closes out this issue. Modulo our API sample code (https://github.com/flutter/flutter/issues/101068), the entire flutter/flutter repository is now migrated to use super parameters and the `use_super_parameters` lint is enforced. Lots of lines of boilerplate code were deleted during the migration and I think it cleans up the code nicely.

Thanks to @pq @bwilkerson @chloestefantsova @leafpetersen @scheglov @johnniwinther @mit-mit @asashour @eernstg @annagrin @gspencergoog @Piinks @christopherfujino @bkonyi for the help in getting this over the line - special thanks for the speedy upstream fixes to resolve the encountered issues and unblock the migration so quickly!
